File: frontend/src/api/tauri/text.ts
/** Text utilities — chunking, page classification, structured extraction. */

import { invoke } from '@tauri-apps/api/core'
import { invokeWithError } from './_utils'
import { ErrorCode } from '../../utils/errors'

// ---- text_ops ----

export interface TextChunkResult {
  chunks: string[]
  total_chunks: number
}

export async function textChunk(text: string, chunkSize = 512, overlap = 128): Promise<TextChunkResult> {
  return invokeWithError(
    () => invoke<TextChunkResult>('text_chunk', { text, chunkSize, overlap }),
    ErrorCode.ApiError,
  )
}

// ---- classifier ----

export interface PageClassification {
  page_idx: number
  text_density: number
  is_scanned: boolean
  has_molecular_patterns: boolean
}

export interface DocumentClassification {
  text_density: number
  is_scanned: boolean
  has_molecular_patterns: boolean
  metadata_hints: Record<string, unknown> | null
  pages: PageClassification[]
  needs_confirmation: boolean
}

export async function classifyPage(pageText: string, pageIdx: number): Promise<PageClassification> {
  return invokeWithError(
    () => invoke<PageClassification>('classify_page', { pageText, pageIdx }),
    ErrorCode.ApiError,
  )
}

export async function classifyDocument(pages: string[], metadata?: Record<string, unknown>): Promise<DocumentClassification> {
  return invokeWithError(
    () => invoke<DocumentClassification>('classify_document', { pages, metadata: metadata ?? null }),
    ErrorCode.ApiError,
  )
}

// ---- extractor ----

export interface ActivityData {
  activity_type: string
  value: number
  units: string
  context: string
}

export async function extractSmilesCandidates(text: string): Promise<string[]> {
  return invoke<string[]>('extract_esmiles_candidates', { text })
}

export async function extractActivities(text: string): Promise<ActivityData[]> {
  return invoke<ActivityData[]>('extract_activities', { text })
}

export interface AssociatedMolecule {
  esmiles: string
  activity: ActivityData | null
  position: number
  confidence: string
  source_doc: string
}

/** Extract e-smiles + nearby activity (200-char window) for association UI. */
export async function extractAssociatedMolecules(
  text: string,
  sourceDoc: string,
): Promise<AssociatedMolecule[]> {
  return invoke<AssociatedMolecule[]>('extract_associated_molecules', {
    text,
    sourceDoc,
  })
}

File: src/mbforge/routers/text.py
"""Text processing, classification, and extraction endpoints.

Most endpoints in this router are thin stubs that the pipeline runs
in-process; the only endpoint that performs real work is
``/extract/activities``, which is wired to the on-demand
``extract_activities_from_document`` LLM table parser.
"""

from __future__ import annotations

from fastapi import APIRouter

from ..utils.logger import get_logger

router = APIRouter()

logger = get_logger("mbforge.routers.text")


@router.post("/text/chunk")
async def text_chunk(body: dict) -> dict:
    """Text chunking stub."""
    text = body.get("text", "")
    return {"chunks": [text] if text else [], "total_chunks": 1 if text else 0}


@router.post("/classify/page")
async def classify_page(body: dict) -> dict:
    """Page classification stub."""
    return {
        "page_idx": body.get("pageIdx", 0),
        "text_density": 0.0,
        "is_scanned": False,
        "has_molecular_patterns": False,
    }


@router.post("/classify/document")
async def classify_document(body: dict) -> dict:
    """Document classification stub."""
    return {
        "text_density": 0.0,
        "is_scanned": False,
        "has_molecular_patterns": False,
        "metadata_hints": None,
        "pages": [],
        "needs_confirmation": False,
    }


@router.post("/extract/esmiles-candidates")
async def extract_esmiles_candidates(body: dict) -> dict:
    """Extract E-SMILES candidates stub."""
    return []


@router.post("/extract/activities")
async def extract_activities(body: dict) -> list[dict]:
    """Extract activities from text using LLM-backed table parser.

    The pipeline runs the same parser in-process during the
    ``extract_activities`` stage; this endpoint lets the front-end
    trigger an on-demand parse for the SAR panel preview. Failures
    degrade to an empty list so the caller never sees a 500 from a
    missing LLM key or a malformed table.
    """
    text = body.get("text", "")
    if not text:
        return []
    try:
        from ..pipeline.extract_activities import (
            _extract_tables_from_markdown,
            _parse_table_with_llm,
        )

        tables = _extract_tables_from_markdown(text)
        if not tables:
            return []
        results: list[dict] = []
        for table_md, _page_num in tables[:5]:  # Phase 0: cap at first 5 tables
            try:
                records = _parse_table_with_llm(
                    table_md, 0, model="gpt-4o-mini", page_num=None
                )
            except Exception as exc:
                logger.warning("Table parse failed: %s", exc)
                continue
            for rec in records:
                results.append(
                    {
                        "activity_type": rec.activity_type,
                        "value": rec.value,
                        "units": rec.unit,
                        "context": rec.raw_text,
                    }
                )
        return results
    except Exception as exc:
        logger.error("Activity extraction failed: %s", exc)
        return []


@router.post("/extract/associated-molecules")
async def extract_associated_molecules(body: dict) -> dict:
    """Extract associated molecules stub."""
    return []
